### ISSUE TITLE
Add DOMNormalizer/DOMSerializerImpl memory leak prevention

### ISSUE THREAD

**Issue by LukeWoodward** [mjiderhamn/classloader-leak-prevention#36]
The ClassLoaderLeakPreventor library does not handle classloader leaks caused
by the `RuntimeException`s in the static final `abort` fields in the JRE classes
`com.sun.org.apache.xerces.internal.dom.DOMNormalizer` and
`com.sun.org.apache.xml.internal.serialize.DOMSerializerImpl`.

When each of these classes is initialized, its `abort` exception is created.  If a web app class happens to be in the call stack at that point, the web app class cannot be garbage-collected when the web app is undeployed because a static field somewhere in a JRE class has a reference to it.

To reproduce the problem, use the following servlet:

    package com.example;

    import javax.servlet.ServletException;
    import javax.servlet.http.*;
    import javax.xml.parsers.DocumentBuilderFactory;
    import org.w3c.dom.Document;

    public class DOMNormalizerLeakServlet extends HttpServlet {
        protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException {
            try {
                Document document = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
                document.normalizeDocument();
                response.getWriter().write("done");
            }
            catch (Exception e) {
                throw new ServletException(e);
            }
        }
    }

and the following web.xml:

    <?xml version="1.0" encoding="UTF-8"?>
    <web-app xmlns="http://xmlns.jcp.org/xml/ns/javaee"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd"
         version="3.1">
        <listener>
            <listener-class>se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor</listener-class>
        </listener>
        <servlet>
            <servlet-name>test</servlet-name>
            <servlet-class>com.example.DOMNormalizerLeakServlet</servlet-class>
        </servlet>
        <servlet-mapping>
            <servlet-name>test</servlet-name>
            <url-pattern>/index.html</url-pattern>
        </servlet-mapping>   
    </web-app>

Here's a path from a destroyed Tomcat web app classloader to `DOMNormalizer.abort`, courtesy of JVisualVM:
    
    this     - value: org.apache.catalina.loader.WebappClassLoader #2
     <- <classLoader>     - class: com.example.DOMNormalizerLeakServlet, value: org.apache.catalina.loader.WebappClassLoader #2
      <- [2]     - class: java.lang.Object[], value: com.example.DOMNormalizerLeakServlet class DOMNormalizerLeakServlet
       <- [2]     - class: java.lang.Object[], value: java.lang.Object[] #4167
        <- backtrace     - class: java.lang.RuntimeException, value: java.lang.Object[] #4166
         <- abort (sticky class)     - class: com.sun.org.apache.xerces.internal.dom.DOMNormalizer, value: java.lang.RuntimeException #1
    
Similarly, if you replace

                document.normalizeDocument();

in the servlet above with

                DOMImplementationLS implementation = (DOMImplementationLS)document.getImplementation();
                implementation.createLSSerializer().writeToString(document);
                
and add an `import` for `org.w3c.dom.ls.DOMImplementationLS`, then you get another memory leak.  Here's a path from another destroyed Tomcat web app classloader to `DOMSerializerImpl.abort`:

    this     - value: org.apache.catalina.loader.WebappClassLoader #1
     <- <classLoader>     - class: com.example.DOMNormalizerLeakServlet, value: org.apache.catalina.loader.WebappClassLoader #1
      <- [2]     - class: java.lang.Object[], value: com.example.DOMNormalizerLeakServlet class DOMNormalizerLeakServlet
       <- [2]     - class: java.lang.Object[], value: java.lang.Object[] #3416
        <- backtrace     - class: java.lang.RuntimeException, value: java.lang.Object[] #3415
         <- abort (sticky class)     - class: com.sun.org.apache.xml.internal.serialize.DOMSerializerImpl, value: java.lang.RuntimeException #1

**Comment by mjiderhamn** [comment]
Hi Luke and thanks for the report. I believe I have [found a way around this](https://github.com/mjiderhamn/classloader-leak-prevention/commit/86ff98537a10e8791a99998b477b53134f57604a), however I have not been able to reproduce the problem with my test suite, and therefore have not verified the fix. Would you mind building the library from the latest sources, and see if it solves your problem and report back?

Thanks in advance.

**Comment by LukeWoodward** [comment]
The latest version of the library does appear to solve the problem.

I copied the latest version of the ClassLoaderLeakPreventor.java class into the test web app I created to reproduce this problem.  I could no longer reproduce the memory leak with either DOMNormalizer or DOMSerializerImpl.

**Comment by mjiderhamn** [comment]
Thank you for the response! I have now released version 1.15.0 that includes this and should propagate to Maven Central shortly.

```xml
<dependency>
  <groupId>se.jiderhamn</groupId>
  <artifactId>classloader-leak-prevention</artifactId>
  <version>1.15.0</version>
</dependency>
```
